pkg/spade-spike/spade-spike.go

```diff
@@ -0,0 +1,286 @@
+	cidlink "github.com/ipld/go-ipld-prime/linking/cid"
+	"github.com/ipld/go-ipld-prime/traversal"
+	"github.com/libp2p/go-libp2p/core/host"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/libp2p/go-libp2p/core/routing"
+	"golang.org/x/exp/slices"
+
+	_ "github.com/ipld/go-codec-dagpb"
+	ipld "github.com/ipld/go-ipld-prime"
+	_ "github.com/ipld/go-ipld-prime/codec/dagcbor"
+	_ "github.com/ipld/go-ipld-prime/codec/dagjson"
+	_ "github.com/ipld/go-ipld-prime/codec/raw"
+)
+
+var logger = logging.Logger("spade-test")
+
+type SingleContentRouter struct {
+	AddrInfo peer.AddrInfo
+}
+
+func (s SingleContentRouter) PutValue(context.Context, string, []byte, ...routing.Option) error {
+	return routing.ErrNotSupported
+}
+
+func (s SingleContentRouter) GetValue(context.Context, string, ...routing.Option) ([]byte, error) {
+	return nil, routing.ErrNotFound
+}
+
+func (s SingleContentRouter) SearchValue(ctx context.Context, key string, opts ...routing.Option) (
+	<-chan []byte, error) {
+	return nil, routing.ErrNotFound
+}
+
+func (s SingleContentRouter) Provide(context.Context, goCid.Cid, bool) error {
+	return routing.ErrNotSupported
+}
+
+func (s SingleContentRouter) FindProvidersAsync(context.Context, goCid.Cid, int) <-chan peer.AddrInfo {
+	ch := make(chan peer.AddrInfo)
+	go func() {
+		ch <- s.AddrInfo
+		close(ch)
+	}()
+	return ch
+}
+
+func (s SingleContentRouter) FindPeer(context.Context, peer.ID) (peer.AddrInfo, error) {
+	return peer.AddrInfo{}, routing.ErrNotFound
+}
+
+func (s SingleContentRouter) Bootstrap(context.Context) error {
+	return nil
+}
+
+func (s SingleContentRouter) Close() error {
+	return nil
+}
+
+type MessageReceiver struct {
+	BSClient       *bsclient.Client
+	MessageHandler func(ctx context.Context, sender peer.ID, incoming bsmsg.BitSwapMessage)
+}
+
+func (m MessageReceiver) ReceiveMessage(
+	ctx context.Context,
+	sender peer.ID,
+	incoming bsmsg.BitSwapMessage) {
+	m.BSClient.ReceiveMessage(ctx, sender, incoming)
+	m.MessageHandler(ctx, sender, incoming)
+}
+
+func (m MessageReceiver) ReceiveError(err error) {
+	m.BSClient.ReceiveError(err)
+}
+
+func (m MessageReceiver) PeerConnected(id peer.ID) {
+	m.BSClient.PeerConnected(id)
+}
+func (m MessageReceiver) PeerDisconnected(id peer.ID) {
+	m.BSClient.PeerDisconnected(id)
+}
+
+type BitswapClient struct {
+	host    host.Host
+	timeout time.Duration
+}
+
+func NewBitswapClient(host host.Host, timeout time.Duration) BitswapClient {
+	return BitswapClient{
+		host:    host,
+		timeout: timeout,
+	}
+}
+
+// Attempts to decode the block data into a node and return its links
+func FindLinks(ctx context.Context, blk blocks.Block) ([]datamodel.Link, error) {
+	if blk.Cid().Prefix().Codec == goCid.Raw {
+		// This can happen at the bottom of the tree
+		return nil, errors.New("raw block encountered " + blk.Cid().String())
+	}
+
+	decoder, err := cidlink.DefaultLinkSystem().DecoderChooser(cidlink.Link{Cid: blk.Cid()})
+
+	if err != nil {
+		return nil, err
```

Comment:
I recommend using errors.Wrap or errors.WithStack to preserve the stacktrace
https://github.com/cockroachdb/errors#how-to-use